### PR TITLE
TRUNK-4253:Made PatientService.savePatient threadsafe

### PR DESCRIPTION
**Pull request by samuelmale**
Ticket: https://issues.openmrs.org/browse/TRUNK-4253

### CODE REVIEW

File: api/src/test/java/org/openmrs/api/PatientServiceTest.java [hunk 1]
```diff
@@ -1848,6 +1853,48 @@ public void savePatient_shouldUpdateAnExistingPatient() throws Exception {
...
+		       
+		Runnable r2 = () -> savePatientOnThread(ctx, patient2);
+				
+		Thread t1 = new Thread(r1);
+		Thread t2 = new Thread(r2);
+		
+		t1.start();
+		t2.start();
+		
+		assertTrue(isSavePatientThreadsafe.get());
```
Comment (suthagar23) [inline api/src/test/java/org/openmrs/api/PatientServiceTest.java]: You should check for `assertFalse` since the default value is set to `false`.

File: api/src/test/java/org/openmrs/api/PatientServiceTest.java [hunk 2]
```diff
@@ -1848,6 +1853,48 @@ public void savePatient_shouldUpdateAnExistingPatient() throws Exception {
 		assertTrue("The gender should be new", patient2.getGender().equals("F"));
 	}
 	
+	@Test
+	public void savePatient_shouldFailSaveIfPatientsWithSameIdentifierAreSaved() {
+	    isSavePatientThreadsafe = new AtomicBoolean(false);
```
Comment (suthagar23) [inline api/src/test/java/org/openmrs/api/PatientServiceTest.java]: Just one opinion,
`isSavePatientThreadsafe` would be changed to `true` by default or `isSavePatientThreadUnSafe = false` by default will good for the naming. Because at the beginning all threads are in safe state and it will become the critical stage while you are performing the saving method.

**Comment by dkayiwa** [comment]
@samuelmale does the unit test fail when you remove the changes in PatientServiceImpl?

**Comment by samuelmale** [comment]
Yes. The test fails @dkayiwa

File: api/src/test/java/org/openmrs/api/PatientServiceTest.java [hunk 3]
```diff
@@ -1881,7 +1934,7 @@ public void getAllPatients_shouldFetchVoidedPatientsWhenGivenIncludeVoidedIsTrue
 		List<Patient> allPatients = patientService.getAllPatients(true);
 		// there are 1 voided and 4 nonvoided patients in
 		// standardTestDataset.xml
-		assertEquals(6, allPatients.size());
+		assertEquals(7, allPatients.size());
```
Comment (wluyima) [inline api/src/test/java/org/openmrs/api/PatientServiceTest.java]: Why is this necessary?

Comment (samuelmale) [reply api/src/test/java/org/openmrs/api/PatientServiceTest.java]: Hi, <br/>I have been facing this challenge for a while. Its all about Threads saving the data to the "in-memory-database" and effects aren't reverted as the default behavior is to other tests. Could I be missing something when using Threads?   
The above happens because of this [line](https://github.com/samuelmale/openmrs-core/blob/ce0953e35f0ddfaff07c738156cbcac386ab82af/api/src/test/java/org/openmrs/api/PatientServiceTest.java#L1876). Meaning the patient is saved the effect ain't reverted hence I had to change the number of patients from 6 to 7.  Its exactly the same issue in this thread. Could you please advise me on this?

@dkayiwa also asked me about it [here](https://github.com/openmrs/openmrs-core/pull/2562#discussion_r177600261).

Comment (wluyima) [reply api/src/test/java/org/openmrs/api/PatientServiceTest.java]: All I know is that changes in one thread aren't visible in others in tests since the transactions get rollback.

What I don't understand is why it's happening now and not before because the test was passing before your changes

**Comment by samuelmale** [comment]
@wluyima  I have added some changes

File: api/src/test/java/org/openmrs/api/PatientServiceTest.java [hunk 4]
```diff
@@ -1848,6 +1852,55 @@ public void savePatient_shouldUpdateAnExistingPatient() throws Exception {
...
+				
+		Thread t1 = new Thread(r1);
+		Thread t2 = new Thread(r2);
+		
+		t1.start();
+		t2.start();
+		
+		try {
+			t1.join();
+			//t2.join();
```
Comment (dkayiwa) [inline api/src/test/java/org/openmrs/api/PatientServiceTest.java]: Did you take a look at this? https://wiki.openmrs.org/display/docs/Java+Conventions#JavaConventions-Donotcommentoutcode

**Comment by samuelmale** [comment]
Just removed the comment @dkayiwa

File: api/src/main/java/org/openmrs/api/impl/PatientServiceImpl.java [hunk 5]
```diff
@@ -116,14 +117,16 @@ public void onShutdown() {
 	/**
 	 * @see org.openmrs.api.PatientService#savePatient(org.openmrs.Patient)
 	 */
+	@Transactional(isolation = Isolation.SERIALIZABLE)
 	@Override
 	public Patient savePatient(Patient patient) throws APIException {
+		synchronized(this) {
 		requireAppropriatePatientModificationPrivilege(patient);
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/api/impl/PatientServiceImpl.java]: Did you intentionally decide not to TAB the code within the synchronised phrase?

File: api/src/test/java/org/openmrs/api/PatientServiceTest.java [hunk 6]
```diff
@@ -1848,6 +1852,56 @@ public void savePatient_shouldUpdateAnExistingPatient() throws Exception {
...
+		}	
+		
+		assertTrue("IdentifierNotUniqueException not thrown, method not Threadsafe", isSavePatientThreadsafe.get());
+		
+		//Cleans up the unrolled back Thread transactions.
+		deleteAllData();
+	}
+	public void savePatientOnThread(UserContext ctx, Patient patient) {
+		Context.setUserContext(ctx);
+        Context.openSessionWithCurrentUser();
```
Comment (dkayiwa) [inline api/src/test/java/org/openmrs/api/PatientServiceTest.java]: Can you also fix the above indention?

Comment (samuelmale) [reply api/src/test/java/org/openmrs/api/PatientServiceTest.java]: I wonder, but locally, the indention is fine. Just to prove, hit view full file and check.

I have highlighted the it here https://github.com/samuelmale/openmrs-core/blob/fd2aed2031b647f046562da4142851669c3c4061/api/src/test/java/org/openmrs/api/PatientServiceTest.java#L1890-L1894

File: api/src/main/java/org/openmrs/api/impl/PatientServiceImpl.java [hunk 7]
```diff
@@ -116,23 +117,23 @@ public void onShutdown() {
...
-		}
-
-		if (!patient.getVoided()) {
-			checkPatientIdentifiers(patient);
-		}
-
-		setPreferredPatientIdentifier(patient);
-		setPreferredPatientName(patient);
-		setPreferredPatientAddress(patient);
+		synchronized(this) {
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/api/impl/PatientServiceImpl.java]: Do you have a reason for using a synchronised block instead of method?

File: api/src/test/java/org/openmrs/api/PatientServiceTest.java (hunk 2 above)
Comment (dkayiwa) [reply api/src/test/java/org/openmrs/api/PatientServiceTest.java]: @samuelmale did you see the above?

**Comment by samuelmale** [comment]
I'm of the view that we use synchronized blocks, a sync'd method makes it not thread safe @dkayiwa 

check the CI failure

**Comment by dkayiwa** [comment]
Can you share the link where you read that?

**Comment by samuelmale** [comment]
I have just run 'mvn clean install' on this branch twice and it builds. I wonder whats happening with CI

@dkayiwa

**Comment by dkayiwa** [comment]
I have restarted the build

File: api/src/test/java/org/openmrs/api/PatientServiceTest.java (hunk 2 above)
Comment (dkayiwa) [reply api/src/test/java/org/openmrs/api/PatientServiceTest.java]: @samuelmale did you see the above?

**Comment by samuelmale** [comment]
Cool, just waiting for yr review @dkayiwa

File: api/src/test/java/org/openmrs/api/PatientServiceTest.java (hunk 6 above)
Comment (samuelmale) [reply api/src/test/java/org/openmrs/api/PatientServiceTest.java]: If this is the unaddressed comment, I'm pretty sure my indention is fine. Have you looked at it as a file. In otherwise have looked at the above link? @dkayiwa

Comment (samuelmale) [reply api/src/test/java/org/openmrs/api/PatientServiceTest.java]: If this is the unaddressed comment, I'm pretty sure my indention is fine. Have you looked at it as a file. In otherwise have looked at the above link? @dkayiwa

**Comment by samuelmale** [comment]
Could you restart the CI build once more @dkayiwa ?

**Comment by dkayiwa** [comment]
Even compiling it locally on your machine should fail too.

**Comment by samuelmale** [comment]
Noo, it actually passes @Kayiwa Daniel <danielkayiwa@gmail.com>

**Comment by dkayiwa** [comment]
@samuelmale i have restarted it several times and noticed that some times it fails while other times it passed. Whenever it fails, the error message is as below:

Results :
Failed tests: 
  PatientServiceTest.getAllPatients_shouldFetchVoidedPatientsWhenGivenIncludeVoidedIsTrue:1938 expected:<6> but was:<7>
  PatientServiceTest.savePatient_shouldFailSaveIfPatientsWithSameIdentifierAreSaved:1884 IdentifierNotUniqueException not thrown, method not Threadsafe
Tests run: 4022, Failures: 2, Errors: 0, Skipped: 36

**Comment by dkayiwa** [comment]
I just run it another time and got this:

Results :
Failed tests: 
  PatientServiceTest.getAllPatients_shouldFetchVoidedPatientsWhenGivenIncludeVoidedIsTrue:1938 expected:<6> but was:<8>
  PatientServiceTest.savePatient_shouldFailSaveIfPatientsWithSameIdentifierAreSaved:1884 IdentifierNotUniqueException not thrown, method not Threadsafe

**Comment by samuelmale** [comment]
Do you mind fetching this branch and testing it out locally?

**Comment by dkayiwa** [comment]
Like i said, it is random. I have just rerun it and now you can see the failure.

**Comment by samuelmale** [comment]
So whats really going on, but locally its not random @dkayiwa

**Comment by dkayiwa** [comment]
Tru dat!
Looks like it depends on the processing power.

**Comment by samuelmale** [comment]
So whats up @dkayiwa

**Comment by dkayiwa** [comment]
@samuelmale the very fact that the test randomly fails in some instances, makes me feel like it needs some more work.

**Comment by samuelmale** [comment]
This weird behaviour started when I changed from a blocked synchronized block to a method synchronized block @dkayiwa . Do we have reasons as to why we can't use the block synchronization?

**Comment by dkayiwa** [comment]
I remember as if it still failed even with block synchronisation. Any way, you can change it back and will show you by simply running travis again.

**Comment by samuelmale** [comment]
Hahaha okie @dkayiwa

**Comment by samuelmale** [comment]
Whats going on here @dkayiwa .
I has failed again. Could you restart travis?

**Comment by samuelmale** [comment]
ahhhhh!!!! But have you ever hit upon such a case? @dkayiwa

**Comment by suthagar23** [comment]
@samuelmale any more ideas on the above Travis failures?

**Comment by bejelith** [comment]
I think ur problem is that savePatient is not thread safe and you have no idea if ur executing inside an pre-existing transaction. Create a separate hibernate transaction/session and should work
